### PR TITLE
Issue #264

### PR DESCRIPTION
**Pull request by DavidMcCann-NOC**
Changes to address Issue #264 causing crashes of find_minima in the case of out-of-bounds x0. 

In test cases with un-physical RSV and Sigma0 this fix does not always flag bad boundary conditions, however it does prevent the function from crashing during parallel computation.

### CODE REVIEW

File: seastar/retrieval/cost_function.py [hunk 1]
```diff
@@ -142,6 +149,40 @@ def find_minima(level1_pixel, noise_pixel, gmf):
...
+        Dict output from scipy.optimize.least_squares.
+
+    Returns
+    -------
+    lmout_null : ``dict``
+        copy of `lmout_dict` with variables changed to NaN.
+
+    """
+    lmout_null = lmout_dict.copy()
+    for var in lmout_dict.keys():
```
Comment (ACHMartin) [inline seastar/retrieval/cost_function.py:174]: very nice, very compact. You're a coding in a nice pythonic way now :)

File: seastar/retrieval/cost_function.py [hunk 2]
```diff
@@ -142,6 +149,40 @@ def find_minima(level1_pixel, noise_pixel, gmf):
...
+        if isinstance(lmout_dict[var], np.ndarray):
+            lmout_null[var] = np.full(lmout_dict[var].shape, np.NaN)
+        else:
+            lmout_null[var] = np.NaN
+
+    lmout_null['message'] = 'x0 found to be infeasible and out of bounds'
```
Comment (ACHMartin) [inline seastar/retrieval/cost_function.py:180]: perfect

**Review by ACHMartin — APPROVED** [review]
I think it will be a good habit to change systematicaly the _version.py. I know it is a bit useless now, as you didn't run anything, So up to you.
Thanks, nice work.

File: seastar/retrieval/cost_function.py [hunk 3]
```diff
@@ -10,6 +10,7 @@
 from scipy.optimize import least_squares
 import seastar
 from seastar.utils.tools import dotdict, da2py
+import sys
```
[the remark this answers is not in the document]
Comment (DavidMcCann-NOC) [reply seastar/retrieval/cost_function.py]: Nope, must have been left in from a test. Deleted now, thanks

File: seastar/retrieval/cost_function.py (hunk 1 above)
Comment (DavidMcCann-NOC) [reply seastar/retrieval/cost_function.py:173]: Come a long way from Matlab!

File: seastar/retrieval/ambiguity_removal.py [hunk 4]
```diff
@@ -77,7 +77,8 @@ def ambiguity_sort_by_cost(lmout):
     index : ``xarray.DataArray``
         a dataArray index. Can be applied directly to lmout with lmout.isel(Ambiguities=index_cost)
     """
-    index = lmout.cost.argmin(dim='Ambiguities')
```
[the remark this answers is not in the document]
Comment (DavidMcCann-NOC) [reply seastar/retrieval/ambiguity_removal.py:80]: Ah, I didn't realise that was an available option. Probably better to use that, thanks

**Comment by DavidMcCann-NOC** [comment]
> > .
> 
> I assume you did some test. Any chance you could push them as unit_test? If too complicated to do it rapidly (half a day), I suggest you put your code for the test in a new issue and we will clean that up by then

I'm not sure I understand - do you mean push my testing notebook for this issue but on the unit_test branch?

**Comment by ACHMartin** [comment]
> > > .
> > 
> > 
> > I assume you did some test. Any chance you could push them as unit_test? If too complicated to do it rapidly (half a day), I suggest you put your code for the test in a new issue and we will clean that up by then
> 
> I'm not sure I understand - do you mean push my testing notebook for this issue but on the unit_test branch?

We have unit test in "tests/...". It will be good to add the tests you have done (quick test in a matter of seconds or less than 1 minute) in the unit_test, so we can test it systematically.
If you don't have time to do it now, raise an issue with the link to your notebook

**Comment by DavidMcCann-NOC** [comment]
> > > > .
> > > 
> > > 
> > > I assume you did some test. Any chance you could push them as unit_test? If too complicated to do it rapidly (half a day), I suggest you put your code for the test in a new issue and we will clean that up by then
> > 
> > 
> > I'm not sure I understand - do you mean push my testing notebook for this issue but on the unit_test branch?
> 
> We have unit test in "tests/...". It will be good to add the tests you have done (quick test in a matter of seconds or less than 1 minute) in the unit_test, so we can test it systematically. If you don't have time to do it now, raise an issue with the link to your notebook

This might take more time than that as there is an open unit_test issue (Issue #259) that potentially needs to be fixed first

**Comment by ACHMartin** [comment]
> > > > > .
> > > > 
> > > > 
> > > > I assume you did some test. Any chance you could push them as unit_test? If too complicated to do it rapidly (half a day), I suggest you put your code for the test in a new issue and we will clean that up by then
> > > 
> > > 
> > > I'm not sure I understand - do you mean push my testing notebook for this issue but on the unit_test branch?
> > 
> > 
> > We have unit test in "tests/...". It will be good to add the tests you have done (quick test in a matter of seconds or less than 1 minute) in the unit_test, so we can test it systematically. If you don't have time to do it now, raise an issue with the link to your notebook
> 
> This might take more time than that as there is an open unit_test issue (Issue #259) that potentially needs to be fixed first

I think this can be done despite of this. Pytest can fail on different items (unit test), it doesn't prevent to add a new unit test and running it (I think).
Issue #259 is not very clear to me.  Anyway, I think you don't have much time, but it will be a pity to lose the testing you have done. Keeping it in a new issue with the associated notebooks will certainly help to do the unit test for the try/catch in the future.
